### PR TITLE
Add support for common collections.abc types

### PR DESCRIPTION
**Pull request by jrmccluskey**
Adds support for the `collections.abc` module's `Iterable`, `Iterator`, and `Generator` containers as type hints. The `typing` module equivalents had previously been supported but are deprecated as of Python 3.9, so support for the new recommended container types was necessary. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by jrmccluskey** [comment]
R: @tvalentyn

**Comment by tvalentyn** [comment]
PTAL at test failures.

```
>         self.assertEqual(converted_beam_type, expected_beam_type, description)
E         AssertionError: collections.abc.Iterable[int] != Iterable[<class 'int'>] : collection iterable
```

**Comment by jrmccluskey** [comment]
Looks like the mechanic I was leveraging to check if the type was an iterable was new behavior in 3.11, not older versions. The new routing of checking for an `__iter__` attribute should fix that.

**Comment by jrmccluskey** [comment]
Run Python_Integration PreCommit

**Comment by tvalentyn** [comment]
> TIL our python codecov is not running on 3.9 or 3.10. Interesting.

Yes, some tests only run on one of the Python versions, typically the lowest.

File: sdks/python/apache_beam/typehints/native_type_compatibility.py [hunk 1]
```diff
@@ -187,6 +187,26 @@ def convert_builtin_to_typing(typ):
...
+  """Converts a given collections.abc type to a typing object.
+
+  Args:
+    typ: an object inheriting from a collections.abc object
+
+  Returns:
+    type: The corresponding typing object.
+  """
+  if hasattr(typ, '__iter__'):
+    print("here")
```
Comment (tvalentyn) [inline sdks/python/apache_beam/typehints/native_type_compatibility.py:200]: leftover

**Review by tvalentyn — APPROVED** [review]

File: sdks/python/apache_beam/typehints/native_type_compatibility.py (hunk 1 above)
Comment (jrmccluskey) [reply sdks/python/apache_beam/typehints/native_type_compatibility.py]: fixed

**Comment by jrmccluskey** [comment]
The MacOS 3.10 unit test run appears to be segfaulting at head, not this PR

**Pull request merged**